File: ngram_models/Python_Jeremy/NGram.py
class NGram:
	items = []
	count = 0
	counter = 0
	
	def __init__(self, size):
		self.count = size
		self.items = [None] * size
		
	def addWord(self, word):
		if self.count > 0:
			self.items[len(self.items)-self.count] = word
			self.count -= 1
			self.counter += 1
			
	def __eq__(self, other):
		if isinstance(other, self.__class__):
			return self.items == other.items
		else:
			return false
			
	def __str__(self):
		if self.counter > 0:
			return str(len(self.items)) + '-GRAM\t' + ' '.join(self.items)
		else: 
			return ""

	def __hash__(self):
		if self.count == 0:
			return hash(' '.join(self.items))
		else:
			return hash("");

File: ngram_models/Java_Jeremy/NGram.java
public class NGram {

	public String[] items;
	private int count;

	public NGram(int size) {
		count = size;
		items = new String[size];
	}

	public void addWord(String word) {
		if(count > 0) {
			items[items.length-count] = word;
			count--;
		}	
	}
	
	public String toString() {
		String output = items.length + "-GRAM\t";
        output += items[0];
		for(int i = 1; i < items.length; i++)
			output += " " + items[i];	
		return output;	
	}
	
	public boolean equals(Object o) {
		String[] strArr = ((NGram)o).items;
		if(strArr.length == items.length) {
			for(int i = 0; i < items.length; i++)
				if(!strArr[i].equals(items[i]))
					return false;
			return true;		
		}
		return false;
	}
	
	public int hashCode() {
		String hashStr = "";
		for(int i = 0; i < items.length; i++)
			hashStr += items[i] + ":";
		int hash = 7;
		for(int i = 0; i < hashStr.length(); i++)
			hash = hash*31+hashStr.charAt(i);
		return hash;
	}
}	
